internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	lokiClientTimeout = 10 * time.Second
+)
+
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+// Creates a new logClient for Loki.
+// The provided URL is expected to be the fully qualified API path, e.g.
+//
+//	https://202.61.12.93:3100/loki/api/v1/push
+func newLokiClient(url string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{
+			Timeout: lokiClientTimeout,
+		},
+		url: url,
+	}
+}
+
+func (c *lokiClient) Close() error {
+	c.client.CloseIdleConnections()
+	return nil
+}
+
+func (c *lokiClient) Send(entries []servicelog.Entry) error {
+	if len(entries) == 0 {
+		return nil // no-op
+	}
+
+	encodedEntries := make([]lokiMessage, 0, len(entries))
+	for _, entry := range entries {
+		encodedEntries = append(encodedEntries, encodeEntry(entry))
+	}
+
+	requestStruct := lokiRequest{
+		Streams: []lokiStream{{
+			// Add default labels
+			// TODO: allow specifying custom labels
+			Labels: map[string]string{
+				"pebble_service": entries[0].Service, // all entries should be from same service
```

Comment:
I still think we should probably drop this from this PR, because otherwise there's this in-between state with one label, and this might get into Juju, and then people will start relying on this `pebble_service` label (which may end up with a different name when we add labels properly in the follow-up). Thoughts?

Reply:
We need to keep it for now. This is a fundamental piece of information which will be lost if we don't have the label. IMO this is needed to make Loki support usable for now.

I would even support having `pebble_service` as a mandatory label for Loki, but this needs to be discussed further.

Reply:
Okay, I guess it's not a big deal; I just didn't want folks to start using this, set up monitoring/filtering based on this tag, and then we change it when we add proper labelling. But let's leave it for now.